### PR TITLE
Add error codes to our custom errors

### PR DESCRIPTION
**Pull request by FabioRosado**
This PR adds error codes and documentation for our custom errors.  There is still a few items left to do on the PR:

- [ ] ~~Add test to confirm that docs have all the error types~~ (can't read this file)
- [x] Handle `fetch` errors better

On the 'handle fetch errors better' front, I am planning to create a custom function that will call fetch under the hood, but handles the status codes and includes the respective error code so we can replace the `errorCode` type from string to ErrorCode similar to `UserError`

Closes: #956

### CODE REVIEW

File: pyscriptjs/src/plugins/fetch.ts [hunk 1]
```diff
@@ -29,7 +32,10 @@ export function calculatePaths(fetch_cfg: FetchConfig[]) {
             fetchPaths.push(from);
             const filename = to_file || from.split('/').pop();
             if (filename === '') {
-                throw new UserError(`Couldn't determine the filename from the path ${from}, supply ${to_file} parameter!`);
+                throw new UserError(
+                    UserError.ErrorCode.FETCH_PARAMETER_ERROR,
+                    `Couldn't determine the filename from the path ${from}, supply ${to_file} parameter!`
```
Comment (madhur-tandon) [inline pyscriptjs/src/plugins/fetch.ts:37]: I guess this is carried from a previous mistake but `${to_file} parameter` should be `"to_file" parameter` since we don't want to use string interpolation here i.e. we don't want to use the value of `to_file` inside the message, but instead communicate that you need to use this.

File: pyscriptjs/tests/unit/pyodide.test.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import type { AppConfig } from '../../src/pyconfig';
 import { Runtime } from '../../src/runtime';
 import { PyodideRuntime } from '../../src/pyodide';
 import { CaptureStdio } from '../../src/stdio';
+import { make_PyScript } from '../../src/components/pyscript';
```
Comment (madhur-tandon) [inline pyscriptjs/tests/unit/pyodide.test.ts:5]: Why is this imported here?

File: pyscriptjs/src/plugins/fetch.ts (hunk 1 above)
Comment (FabioRosado) [reply pyscriptjs/src/plugins/fetch.ts:37]: Ah thanks! I'll fix it in the next push 😄

File: pyscriptjs/tests/unit/pyodide.test.ts (hunk 2 above)
Comment (FabioRosado) [reply pyscriptjs/tests/unit/pyodide.test.ts:5]: Oops this was a mistake (I thought I was poking at it on Jeff's branch but seems not 😬 )

**Comment by antocuni** [comment]
I don't understand: the PR is not marked as WIP nor as draft, but from the first message I seem to understand it's not complete yet. Is it ready for review or I should wait?

**Comment by FabioRosado** [comment]
Yeah sorry this is ready for review, I was trying to read the exception docs, but couldn't get it to work

**Review by marimeireles — APPROVED** [review]
I'm +1 :)
Thank you for your work here!

**Comment by madhur-tandon** [comment]
It seems like there is just one Enum while there are 3 ways to access it -- 

1. `UserError.ErrorCode.BAD_CONFIG`
2. `FetchError.ErrorCode.BAD_CONFIG`
3. `ErrorCode.BAD_CONFIG`

If both `UserError` and `FetchError` share the same Enum, why is it a property of those classes? Maybe using just the 3rd option is sufficient everywhere?

Or perhaps separate enums can be created specific to `UserError` and `FetchError`.
WDYT?

File: docs/reference/exceptions.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+| Error code | Description                    | Recommendation     |
+|------------|--------------------------------|--------------------|
+| PY0001     | Feature is deprecated          | Check the documentation for the feature you are using and see recommendations. |
+| PY1000     | Invalid configuration supplied | Confirm that your `py-config` tag is using a valid `TOML` or `JSON` syntax and is using the correct configuration type. |
+
+
+
+## Fetch Errors
+
+These error codes are related to any exception raised when trying to fetch a resource. If, while trying to fetch a resource, we encounter a status code that is not 200, the error code will contain the HTTP status code and the `PY2` prefix. For example, if we encounter a 404 error, the error code will be `PY2404`.
```
[the remark this answers is not in the document]
Comment (FabioRosado) [reply docs/reference/exceptions.md:17]: I chose the `PY` prefix mostly because that's our logo haha then the number after is referred to the type of error.

The idea is to have something like:
- PY0 (deprecated and dev one which should be changed to a proper error code)
- PY1 (any configuration issue)
- PY2 (any fetch issue)

As we expand the exceptions we could create different buckets so its easier to know where that error occurred from looking at it 😄

**Comment by FabioRosado** [comment]
> > we can do `FetchError.UserError.BAD_THING_D:`
> 
> I guess you meant `FetchError.ErrorCode.BAD_THING_D` here instead of `FetchError.UserError.BAD_THING_D` right?

Yeah sorry, was thinking about the UserError haha I meant `ErrorCode`

> > this is so we don't have to import both FetchError and UserError all over the place.
> 
> Yes but even using `FetchError.ErrorCode.BAD_THING_D` implies having to import `FetchError` while we could just get away with `ErrorCode.BAD_THING_D` I guess.

Importing `ErrorCode` itself won't create the exception, why would we want to import it on its on (unless we are doing a check for the error code itself? 🤔 )

The idea is that our custom exceptions must contain an ErrorCode and you can't raise one without it

**Comment by madhur-tandon** [comment]
I guess what I really mean is the following:

Currently we do this:

```
throw new UserError(UserError.ErrorCode.FETCH_PARAMETER_ERROR, `Cannot use 'to_file' and 'files' parameters together!`);
```

while we could do this instead:

```
throw new UserError(ErrorCode.FETCH_PARAMETER_ERROR, `Cannot use 'to_file' and 'files' parameters together!`);
```

I am just saying that accessing `UserError.ErrorCode.FETCH_PARAMETER_ERROR` is same as `FetchError.ErrorCode.FETCH_PARAMETER_ERROR` which is same as `ErrorCode.FETCH_PARAMETER_ERROR`

so we don't really need `ErrorCode` to be a property of `UserError` or `FetchError`

**Comment by antocuni** [comment]
> so we don't really need `ErrorCode` to be a property of `UserError` or `FetchError`

as @FabioRosado said, this design was chosen to avoid having to import both `UserError` and `ErrorCode` *everywhere*. But when I proposed this solution I didn't think about `FetchError`, and I agree that now it looks a bit confusing.

In particular, it looks confusing because one quickly starts to wonder whether `UserError.ErrorCode` contains the same codes as `FetchError.ErrorCode`, whether they are the same objects, two copies of the same object, whether I can mix the code, etc. etc.
In short: it creates lot of potential confusion just to avoid one import. So +1 remove it from `UserError` and require the user to import it explicitly.

File: pyscriptjs/src/components/pyscript.ts [hunk 4]
```diff
@@ -18,17 +18,14 @@ export function make_PyScript(runtime: Runtime) {
...
-                    );
-                    _createAlertBanner(errorMessage);
+                try {
+                    const response = await fetchIt(url);
+                    return await response.text();
+                } catch(e) {
+                    _createAlertBanner(e.message);
                     this.innerHTML = '';
-                    throw new FetchError(errorMessage);
+                    throw e
```
Comment (antocuni) [inline pyscriptjs/src/components/pyscript.ts:27]: [follow-up]
Another `_createAlertBanner`: this is a code smell ;).
I think that the correct way to handle this case is to raise `UserError` from here, so that it will be caught by the general UserError logic.

We could go even further, by choosing one of these two options:
1. rename it as `FetchUserError` and make it a subclass of `UserError`: this way, *all* fetch errors will be displayed to the user in a proper way. I'm not 100% sure that this is always the case though: maybe there are cases in which a failed fetch is not an user error but an internal error?

2. in case we decide that some are user errors and some are internal errors, we can probably add an option to `fetchIt` to indicate which exception to raise in each case

File: pyscriptjs/src/plugins/fetch.ts [hunk 5]
```diff
@@ -14,7 +14,10 @@ export function calculatePaths(fetch_cfg: FetchConfig[]) {
         {
             if (to_file !== undefined)
             {
-                throw new UserError(`Cannot use 'to_file' and 'files' parameters together!`);
+                throw new UserError(
+                    UserError.ErrorCode.FETCH_PARAMETER_ERROR,
+                    `Cannot use 'to_file' and 'files' parameters together!`
+                );
```
Comment (antocuni) [inline pyscriptjs/src/plugins/fetch.ts:20]: this seems more a `BAD_CONFIG` error, not a fetch error (we didn't even try to fetch anything).
Also, when the fetch logic becomes a "proper" plugin, this should probably be checked inside the `configure` hook (and then it becomes even more evident that it's a `BAD_CONFIG` error)

File: pyscriptjs/src/pyodide.ts [hunk 6]
```diff
@@ -112,10 +111,7 @@ export class PyodideRuntime extends Runtime {
                 this.interpreter.FS.mkdir(eachPath);
             }
         }
-        const response = await fetch(fetch_path);
-        if (response.status !== 200) {
-            throw new FetchError(`Unable to fetch  ${fetch_path}, reason: ${response.status} - ${response.statusText}`);
-        }
+        const response = await fetchIt(fetch_path);
```
Comment (antocuni) [inline pyscriptjs/src/pyodide.ts:114]: I love how this becomes simpler 😍

File: pyscriptjs/src/utils.ts [hunk 7]
```diff
@@ -50,21 +50,21 @@ export function handleFetchError(e: Error, singleFile: string) {
...
             singleFile +
-            `</u> failed with error 404 (File not Found). Are your filename and path are correct?</p>`;
+            `</u> failed with error 404 (File not Found). Are your filename and path are correct?`;
     } else {
-        errorContent = `<p>PyScript encountered an error while loading from file: ${e.message} </p>`;
+        errorContent = `PyScript encountered an error while loading from file: ${e.message}`;
     }
-    throw new UserError(errorContent, "html");
+    throw new UserError(UserError.ErrorCode.FETCH_ERROR, errorContent, "html");
 }
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:68]: [follow-up]
this seems to be related to the comment which I made above about `FetchUserError` & co.
I think that we can improve this messy logic: instead of having a "catch-all" function which tries to interpret the error *after the fact*, we could probably be proactive: every call to `fetchRobust(userError=True)` (pseudocode) should also indicate an user-friendly explanation. 
 
For example, if we fetch something from the `[[fetch]]` config, saying `Check your filename` might be useful. But if we fail to fetch some internal file (e.g. the pyodide stdlib), then saying `Check your filename` is just confusing to the user. But again, this sounds like a topic for a separate discussion

File: pyscriptjs/tests/integration/test_01_basic.py [hunk 8]
```diff
@@ -145,7 +145,13 @@ def test_py_script_src_not_found(self):
             self.check_js_errors()
 
         error_msg = str(exc.value)
-        assert "Failed to fetch" in error_msg
+        if self.is_fake_server:
+            assert "Failed to fetch" in error_msg
```
Comment (antocuni) [inline pyscriptjs/tests/integration/test_01_basic.py:149]: we need a proper fix for the fake server, eventually :(

**Review by antocuni — CHANGES_REQUESTED** [review]
Good job, I like it!

I don't want this PR to become an endless task. I have marked some comments as `[follow-up]`, meaning that they should not prevent this PR to be merged but that they should probably fixed/improved soon.

Let me also put a [follow-up] immediately:
one of the best advantages of using error codes is to be able to provide precise links to the docs. So I would like that when we display an UserError, we also include a link to e.g. `docs.pyscript.net/blablabla/errors.html#PY1234` or something like that.

File: pyscriptjs/src/utils.ts [hunk 9]
```diff
@@ -50,21 +50,21 @@ export function handleFetchError(e: Error, singleFile: string) {
...
+            `PyScript: Loading from file <u>` +
             singleFile +
-            `</u> failed with error 404 (File not Found). Are your filename and path are correct?</p>`;
+            `</u> failed with error 404 (File not Found). Are your filename and path are correct?`;
     } else {
-        errorContent = `<p>PyScript encountered an error while loading from file: ${e.message} </p>`;
+        errorContent = `PyScript encountered an error while loading from file: ${e.message}`;
     }
-    throw new UserError(errorContent, "html");
+    throw new UserError(UserError.ErrorCode.FETCH_ERROR, errorContent, "html");
```
Comment (antocuni) [inline pyscriptjs/src/utils.ts:67]: [follow-up]
this seems to be related to the comment which I made above about FetchUserError & co.
I think that we can improve this messy logic: instead of having a "catch-all" function which tries to interpret the error after the fact, we could probably be proactive: every call to fetchRobust(userError=True) (pseudocode) should also indicate an user-friendly explanation.

For example, if we fetch something from the [[fetch]] config, saying Check your filename might be useful. But if we fail to fetch some internal file (e.g. the pyodide stdlib), then saying Check your filename is just confusing to the user. But again, this sounds like a topic for a separate discussion

File: docs/reference/exceptions.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+| Error Code | Description                                                  |
+|------------|--------------------------------------------------------------|
+| PY2000     | Generic fetch error, failed to fetch page from the server    |
+| PY2001     | Parameter supplied to fetch is invalid                       |
+| PY2002     | Name supplied when trying to fetch resource is invalid       |
+| PY2401     | You are not authorized to access this resource.              |
+| PY2403     | You are not allowed to access this resource.                 |
+| PY2404     | The page you are trying to fetch does not exist.             |
+| PY2500     | The server encountered an internal error.                    |
+| PY2503     | The server is currently unavailable.                         |
```
[the remark this answers is not in the document]
Comment (FabioRosado) [reply docs/reference/exceptions.md:29]: That's a good idea! Thanks

File: docs/reference/exceptions.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Exceptions
+
+When creating pages with PyScript, you may encounter exceptions. Each handled exception will contain a specific code which will give you more information about it.
+This reference guide contains the error codes you might find and a description of each of them.
+
+## User Errors
+
+| Error code | Description                    | Recommendation     |
+|------------|--------------------------------|--------------------|
+| PY0001     | Feature is deprecated          | Check the documentation for the feature you are using and see recommendations. |
```
[the remark this answers is not in the document]
Comment (FabioRosado) [reply docs/reference/exceptions.md:10]: Right, we talked about that. I'll move deprecated things to the 2 bucket and add some example since we have deprecated stuff 😄

File: pyscriptjs/src/components/pyscript.ts (hunk 4 above)
Comment (FabioRosado) [reply pyscriptjs/src/components/pyscript.ts:27]: That's a good point, I like the both ideas 😄 
When I was creating the error codes, I was wondering how often will people see weird error codes like 500's. Maybe we could always assume it's a user error? 🤔

Comment (antocuni) [reply pyscriptjs/src/components/pyscript.ts:28]: do you mean HTTP 500? I think that yes, in general we should have a way to inform the user that something went wrong. But it seems like a topic for a broader discussion, let's not block this PR on this

File: docs/reference/exceptions.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+| Error code | Description                    | Recommendation     |
+|------------|--------------------------------|--------------------|
+| PY1000     | Invalid configuration supplied | Confirm that your `py-config` tag is using a valid `TOML` or `JSON` syntax and is using the correct configuration type. |
+| PY9000     | Top level await is deprecated  | Create a coroutine with your code and schedule it with `asyncio.ensure_future` or similar |
+
+
+
+## Fetch Errors
+
+These error codes are related to any exception raised when trying to fetch a resource. If, while trying to fetch a resource, we encounter a status code that is not 200, the error code will contain the HTTP status code and the `PY2` prefix. For example, if we encounter a 404 error, the error code will be `PY2404`.
```
Comment (antocuni) [inline docs/reference/exceptions.md:17]:
```suggestion
These error codes are related to any exception raised when trying to fetch a resource. If, while trying to fetch a resource, we encounter a status code that is not 200, the error code will contain the HTTP status code and the `PY2` prefix. For example, if we encounter a 404 error, the error code will be `PY0404`.
```

File: pyscriptjs/src/exceptions.ts [hunk 13]
```diff
@@ -2,20 +2,49 @@ const CLOSEBUTTON = `<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'
...
+  FETCH_NAME_ERROR = "PY2002",
+  // Currently these are created depending on error code received from fetching
+  FETCH_UNAUTHORIZED_ERROR = "PY0401",
+  FETCH_FORBIDDEN_ERROR = "PY0403",
+  FETCH_NOT_FOUND_ERROR = "PY0404",
+  FETCH_SERVER_ERROR = "PY0500",
+  FETCH_UNAVAILABLE_ERROR = "PY0503",
+  BAD_CONFIG = "PY1000",
+  TOP_LEVEL_AWAIT = "PY9000"
+}
```
Comment (antocuni) [inline pyscriptjs/src/exceptions.ts:25]: the FETCH_* errors doesn't seem to match what's described in the doc, they still use the old `PY2...` bucket

**Review by antocuni — APPROVED** [review]
LGTM, apart the error numbers which seems do be different in the code and in the docs
